development-practices/defining-the-boundaries-of-a-package.md

```diff
@@ -0,0 +1,107 @@
+
+The first three package principles are about package _**cohesion**_, they tell us what to put inside packages:
+
+| **REP** | The Release Reuse Equivalency Principle | _The granule of **reuse** is the granule of **release**._      |
+| ------- | --------------------------------------- | -------------------------------------------------------------- |
+| **CCP** | The Common Closure Principle            | _Components that **change together** are packaged together._   |
+| **CRP** | The Common Reuse Principle              | _Components that are **used together** are packaged together._ |
+
+\
+The last three principles are about the _**couplings**_ between packages and talk about metrics that evaluate the package structure of a system.\
+
+
+| **ADP** | The Acyclic Dependencies Principle | _The dependency graph of packages must have no cycles._ |
+| ------- | ---------------------------------- | ------------------------------------------------------- |
+| **SDP** | The Stable Dependencies Principle  | _Depend in the direction of stability._                 |
+| **SAP** | The Stable Abstractions Principle  | _Abstractness increases with stability. (Code-related)_ |
+
+
+
+Principles in other discipline are just as applicable to Salesforce as they are to non-Salesforce Platform. The problem domain was well established before Salesforce, and we more recognize they apply now more than ever.
+
+The Anatomy of a Package is important to understand; however, it is also important to apply principles to packages to obviate issues we would encounter later.
+
+### Principles on Package Cohesion
+
+#### The Release Reuse Equivalency Principle (REP)
+
+> **The granule of reuse is the granule of release. Only components that are released through a tracking system can be effectively reused. This granule is the package **
+
+Reusability is one of the most oft claimed goals of OOD. But what is reuse? &#x20;
+
+Reuse can come in different flavours
+
+* Textually inserting a bunch of code/components from one source to another
+* Utilizing an existing module from someone else and brining into the current source
+
+The above examples of Reuse are of code/component copying; and it comes with a serious disadvantage, one must own the code/component that we copied. If the code/component doesn't work in the target environment, it must be changed. If there are bugs, it might be fixed. Fixes from the original author must be manually merged into the source.  Eventually the code/component copied diverges from the original that it can be hardly. While code/component copying can make it easier to do some initial development; it does not help very much with the most expensive phase of the software lifecycle, _**maintenance**_.
```

Comment:
```suggestion
The above examples of Reuse are of code/component copying; and it comes with a serious disadvantage, one must own the code/component that we copied. If the code/component doesn't work in the target environment, it must be changed. If there are bugs, it might be fixed. Fixes from the original author must be manually merged into the source.  Eventually the code/component copied diverges from the original that it can be hardly recognized. While code/component copying can make it easier to do some initial development; it does not help very much with the most expensive phase of the software lifecycle, _**maintenance**_.
```